ca/certificates.go

```diff
@@ -415,6 +390,100 @@ func NewRootCA(certBytes, keyBytes []byte, certExpiry time.Duration) (RootCA, er
 	return RootCA{Signer: signer, Key: keyBytes, Digest: digest, Cert: certBytes, Pool: pool}, nil
 }
 
+// ValidateCertChain checks that that a bundle certificates for a chain, each one certifying the one above,
+// as per RFC5246 section 7.4.2. They must also form at least one complete chain up to the root pool provided.
+// If everything validates, the entire parsed chain is returned.  Note that the chain can contain some unused
+// certificates (ones that aren't used to form a chain up to the given root pool).  This validation also enforces
+// that none of the certificates in the chain can be expired (unless allowExpired is set to true) or not yet
+// valid. This always returns at least 1 certificate.
+func ValidateCertChain(rootPool *x509.CertPool, certs []byte, allowExpired bool) ([]*x509.Certificate, error) {
+	// Parse all the certificates in the cert bundle
+	parsedCerts, err := helpers.ParseCertificatesPEM(certs)
+	if err != nil {
+		return nil, err
+	}
+	if len(parsedCerts) == 0 {
+		return nil, errors.New("no certificates to validate")
+	}
+	now := time.Now()
+	// ensure that they form a chain, each one being signed by the one after it
+	var intermediatePool *x509.CertPool
+	for i, cert := range parsedCerts {
+		// Manual expiry validation because we want more information on which certificate is expired, and because
+		// we want the option to allow expired certs.
+		if now.Before(cert.NotBefore) {
+			return nil, errors.Wrapf(
+				x509.CertificateInvalidError{
+					Cert:   cert,
+					Reason: x509.Expired,
+				},
+				"certificate (%d - %s) not valid before %s, and it is currently %s",
+				i+1, cert.Subject.CommonName, cert.NotBefore.UTC().Format(time.RFC1123), now.Format(time.RFC1123))
+		}
+		if !allowExpired && now.After(cert.NotAfter) {
+			return nil, errors.Wrapf(
+				x509.CertificateInvalidError{
+					Cert:   cert,
+					Reason: x509.Expired,
+				},
+				"certificate (%d - %s) not valid after %s, and it is currently %s",
+				i+1, cert.Subject.CommonName, cert.NotAfter.UTC().Format(time.RFC1123), now.Format(time.RFC1123))
+		}
+
+		if i > 0 {
+			// check that the previous cert was signed by this cert
```

Comment:
Yep I am totally not a cryptographer.  :)  Go seems pretty open in what they accept though for certificate order, so long as it can establish some chain.  When we parse everything in, it will re-order the certificates correctly.  Am trying to see whether openssl or other TLS clients accept out-of-order certificates.  The spec doesn't seem to have a specific error type or code for out-of-order certs, so I'm not sure who else enforces only accepting this exact order.